### ISSUE TITLE
Optional configuration for disallowing the int/float/complex special case

### ISSUE THREAD

**Issue by robtovey** [microsoft/pyright#10274]
There is a special case in the typing specification that instances of `int` should not be reported as typing errors when used in place of a value marked as `float` or `complex`. This is well-known to exist and well-known often to cause confusion:
- [current documentation](https://typing.python.org/en/latest/spec/special-types.html#special-cases-for-float-and-complex)
- [PEP484](https://peps.python.org/pep-0484/#the-numeric-tower) where it was introduced
- a [python forum post](https://discuss.python.org/t/clarifying-the-float-int-complex-special-case/54018) clarifying the behaviour
- [comment on pyright issue](https://github.com/microsoft/pyright/issues/1575#issuecomment-956377784) clarifying that current behaviour is based on PEPs and typeshed definitions
- [example](https://github.com/microsoft/pyright/issues/7449) of behaviour in bug report

I think the forum post linked above motivates the confusion well. If it helps to put a personal perspective here, there are two sources of bugs I would like to stamp out:
- we use a third-party library with no type hints which has different behaviour for `int` and `float`, mainly because it interfaces with a C library
- internally we have types like `None | float | tuple[float,...]` which people occasionally just narrow with `if isinstance(x, float)` and the `int` case slides through with `None`. An easy bug to fix when identified, but this always causes errors at the most inconvenient times...

Pyright has [lots of optional rules](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options) which are stricter than the guidelines. My feature request is that there is a new rule like `strictNumberInference` which will generate/supress diagnostics when a numeric type will not pass the relevant `isinstance` check.

There are many examples, I will be brief in what my expected behaviour for such a check would be:
```python

def relaxed_half(x: int|float) -> float:
    return x / 2

relaxed_half(1.0) # always valid
relaxed_half(1) # always valid, because the type signature explicitly permits int arguments

def strict_half(x: float) -> float:
    assert isinstance(x, float)
    return x / 2

strict_half(1.0) # always valid
strict_half(1) # invalid whenever strictNumberInference is enabled
```

My apologies if this has been discussed and dismissed before, I couldn't find it in past issues, there are certainly edge-cases and implementation details which I'm not aware of. 

Thank you for your time, both in terms of this request, and for the tool you've produced which makes greatly improves my life every day.

**Comment by erictraut** [comment]
I understand the problem you're describing and your feature request, but I don't think this is the right solution to the problem. 

This issue has been discussed at great length in the Python typing forum in [this thread](https://discuss.python.org/t/options-for-a-long-term-fix-of-the-special-case-for-float-int-complex/54289). As part of that discussion, some investigations were done to inform our options, and it became pretty clear that any attempt to redefine `float` to mean "only a `float`" would break almost all major packages (including most of the stdlib typeshed stubs) and existing code bases. So unless your code base is completely isolated and doesn't import any other libraries (including stdlib modules), you would probably find that your suggested proposal would be unusable in practice. This behavioral change would go well beyond type inference. It would affect the way that type annotations are interpreted both by static and runtime type checkers, and it would affect core type concepts like [assignability](https://typing.python.org/en/latest/spec/glossary.html#term-assignable) and [equivalence](https://typing.python.org/en/latest/spec/glossary.html#term-equivalent).

I also don't think that it makes sense for pyright to introduce configuration options that cause it to be non-compliant with the typing spec. Over the past couple of years, the typing community (myself included) have made big investments to clarify and standardize typing behaviors, and I've personally invested to make pyright [conformant](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) with the typing spec. There is no precedent for adding configuration switches to make pyright non-conformant, and I wouldn't want to create such a precedent.

I think this is something that should be fixed in the type system, not "patched" by individual type checkers through configuration options. Many of the solutions that were proposed in previous discussions would be very disruptive — something that the typing council would be unlikely to support. But there are straightforward and relatively cheap solutions with no backward compatibility impact. My preferred approach is to introduce two new special forms that mean "just a `float`" and "just a `complex`". (This is "option 4" in the typing forum thread, BTW.) This approach would leave in place the historical meanings of `float` and `complex` when they appear within a type annotation. As for naming ideas, my vote would be `typing.Float` and `typing.Complex`.

If this is something you're passionate about, you can contribute to the existing discussion in the typing forum. Or you could start a new thread and reference the existing one. The full process for introducing changes and additions to the typing spec are documented [here](https://github.com/python/typing-council#decisions). This one is straightforward enough that it probably wouldn't require a full PEP. 

One additional thing that you may not be aware of... Pyright's type narrowing logic for `isinstance` already has some special casing for `float` and `complex`. If you attempt to narrow a value whose type is annotated as `float` (which is interpreted as `float | int`), then pyright narrows it into its constituent subtypes.

```python
def func(x: None | float):
    if not isinstance(x, float):
        reveal_type(x)  # int | None
```
